File: LeetCode/Medium/Problems 1-50/Problem 19/p19Sol1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        left = head
        right = head
        
        for i in range(n):
            right = right.next
            
        #Remove first node
        if right is None:
            return head.next
               
        while right.next is not None:
            left = left.next
            right = right.next
         
        #Remove middle or end node
        left.next = left.next.next
        return head 

File: LeetCode/Medium/Problems 1-50/Problem 19/p19Sol1.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode* left = head;
    struct ListNode* right = head;
    
    for (int i = 0; i < n; i++) {
        right = right->next;
    }
    
    //Remove first node
    if (right == NULL) {
        return head->next;
    }
    
    while (right->next != NULL) {
        left = left->next;
        right = right->next;
    }
    
    //Remove middle or end node
    left->next = left->next->next;
    return head;
}